### PR TITLE
Fixes #27639 - Allow users to specify requirements.yml

### PR DESCRIPTION
**Pull request by sjha4**
**To Do:**

- [x] Tests
- [x] The uploaded file doesn't reflect soon (working text?) in the requirements field.
- [x] Allow uploading requirements.yml on update page

****************
**To Test:**
1. Run rails db:migrate

**Using only requirements field:**
Create an ansible collection repo with requirements field as:
```
---
collections:
- name: testing.ansible_testing_content
  version: ">=1.0.0,<=2.0.0"
  source: https://galaxy-dev.ansible.com
- testing.k8s_demo_collection

```
Sync.

**Using uploaded file**
Create a file using the requirement yml above.
Create a collection repo and upload this using the upload button below requirement text area.

requirement

Wait for the uploaded text to appear in the requirement field.
Create and sync.

### CODE REVIEW

**Comment by sjha4** [comment]
@jlsherrill : Should be ok to review. I will add a few tests and work on a couple of UI snags if this looks like the way to go.

Screencast from 08-16-2019 04_29_56 PM

**Comment by johnpmitsch** [comment]
I haven't checked this out yet, but some initial thoughts from our convo and the screenshots:

It looks like we are requiring a `requirements.yml` file for every sync from ansible galaxy. I know the `requirements.yml` file is an option for ansible-galaxy and its good that we are supporting it, but now the basic use case requires you to use the `requirements.yml` file.

For example, I want to install the latest `dcos.cluster` collection, I can just:
```
ansible-galaxy install dcos.cluster
```

But in Katello, I have to specify a requirements.yml with a list of collections, something like this (syntax may be off, untested)
```yaml
collections:
- dcos.cluster
```
and then fill that in, which requires the user to 1) know the syntax that is expected 2) input it without any errors

I think we could simplify the basic use case of wanting to sync the latest version of a single collection down.

Could you have a field called `collections` that they can just input a collection (or maybe multiple comma-seperated ones) and then sync that? I haven't looked at the code, but I'm guessing we could build a simple requirements file for them if that is needed.

These are just my suggestions but we can also have a discussion with @rohoover for some UX guidance

**Comment by johnpmitsch** [comment]
Two other things that I think could improve this form:
1) Maybe we could fill in the ansible galaxy URL in the upstream URL box? It can, of course, be overwritten if users have their own hosted collections, but I imagine using galaxy is the larger use case. You could also do a placeholder text for a more subtle suggestion
2) Link to the ansible docs about requirements.yml in the requirements field

**Comment by jlsherrill** [comment]
@johnpmitsch the main use case we're trying to cover here is to sync ansible collections from galaxy to be used from ansible tower/awx.   the idea being that they would likely either want to sync a) all collections in galaxy, b) some subset as defined by a requirements file they already have or c) all collections within an organization in galaxy (not yet implemented in galaxy AFAIK)

What is the use case to sync a single collection?  when would a user want to do that?

**Comment by johnpmitsch** [comment]
> @johnpmitsch the main use case we're trying to cover here is to sync ansible collections from galaxy to be used from ansible tower/awx. the idea being that they would likely either want to sync a) all collections in galaxy, b) some subset as defined by a requirements file they already have or c) all collections within an organization in galaxy (not yet implemented in galaxy AFAIK)
> 
> What is the use case to sync a single collection? when would a user want to do that?

Thanks for clarifying, I wasn't aware of the mirroring case and was just approaching it as someone who is familiar with installing ansible galaxy content.

I wonder if the form could still benefit from a little more verbosity? Can we tell the users that not filling out the requirements will sync all ansible galaxy collections from the upstream URL? This would make the mirroring use case a bit more clear. For example, "Leave this section blank to sync all ansible collections from the upstream URL"

File: app/services/katello/pulp3/repository/ansible_collection.rb [hunk 1]
```diff
@@ -36,7 +36,7 @@ def remote_options
           if root.url.blank?
             super
           else
-            common_remote_options.merge(url: root.url) # whitelist: root.ansible_collection_whitelist)
+            common_remote_options.merge(url: root.url, requirements_file: root.ansible_collection_requirements)
```
Comment (jlsherrill) [inline app/services/katello/pulp3/repository/ansible_collection.rb]: i'm seeing an issue where if i add a requirements file, save it, then remove the requirements file contents and try to save again we end up sending an empty string, causing an error in pulp.  We probably need to check if the requirements contents are blank and send nil if they are.

Comment (sjha4) [reply app/services/katello/pulp3/repository/ansible_collection.rb]: Updated to send nil as requirement string when blank..

**Comment by sjha4** [comment]
New Page:
Creation_page
Upstream URL popover:
Upstream_URL_popover
Requirement.yml popover:
Requirement_Popover
Details Page:
Ansible_info_page

**Comment by johnpmitsch** [comment]
I really like the new layout and the tooltips!

A few suggestions
- Change ` For the entire Galaxy:` to `For all collections in Ansible Galaxy:`
- Mention what to put in the URL box if you use a requirements.yml file
- I need to click on a tool tip to open it, but I think the rest of our app we hover on them to open? (I'm looking for one now to make sure) Can we hover to open for this?

My suggestion for the help menu is:
```
You can sync collections utilizing just a URL without a requirements.yml file:
1. For all collections in Ansible Galaxy:
https://galaxy.ansible.com/api/v2/collections 
2. For a specific collection:
https://galaxy.ansible.com/api/v2/collections/testing/k8s_demo_collection 
3. Filtering: 
https://galaxy.ansible.com/api/v2/collections/?namespace__name=testing

Or you can use the base URL https://galaxy.ansible.com/ and then use a requirements.yml file below to specify collections
```
Maybe they can use a filtered URL and a requirements.yml file too (I can try this out), but we probably want to keep the scenarios we detail out to a reasonable number of the more useful ones.

**Comment by johnpmitsch** [comment]
`https://galaxy.ansible.com/api/v2/collections/?namespace__name=testing` seems to sync all of the collections, not just the ones in the `testing` namespace

**Comment by sjha4** [comment]
> `https://galaxy.ansible.com/api/v2/collections/?namespace__name=testing` seems to sync all of the collections, not just the ones in the `testing` namespace

@johnpmitsch : Yes..There's an open issue https://github.com/ansible/galaxy/issues/1974 on galaxy for this.

**Comment by sjha4** [comment]
Updated help text: 
Updated_URL

**Comment by sjha4** [comment]
Mirroring text change:
Updated_URL

**Comment by sjha4** [comment]
Updated_URL

**Comment by sjha4** [comment]
> but the upstream url can't end in `/` (returns an error) and the example urls do. So either the example needs to change or we chomp off the "/" before handing to pulp because the example given does error as-is.

@johnpmitsch : Added change to handle trailing "/" . We had an [issue](https://pulp.plan.io/issues/4996) open in pulp around this and an error message was added around this but still disallowing the trailing "/".

**Comment by johnpmitsch** [comment]
@sjha4 I'm getting another error syncing a repository syncing 
```yaml
collections:
# With just the collection name
- crivetimihai.virtualization
```
with upstream URL: `https://galaxy.ansible.com/`

`Response body: {"base_path":["This field must be unique."]}`

Is this a valid repo?

Full tracebacks and info:
```
Actions::Pulp3::Repository::RefreshDistribution
Input:
{"repository_id"=>30,
 "smart_proxy_id"=>1,
 "contents_changed"=>true,
 "remote_user"=>"admin",
 "remote_cp_user"=>"admin",
 "locale"=>"en",
 "current_request_id"=>nil,
 "current_timezone"=>"America/New_York",
 "current_user_id"=>4,
 "current_organization_id"=>1,
 "current_location_id"=>2}
Output:
{}
Exception:
PulpAnsibleClient::ApiError: Error message: the server returns an error
HTTP status code: 400
Response headers: {"date"=>"Fri, 30 Aug 2019 13:11:45 GMT", "server"=>"gunicorn/19.9.0", "content-type"=>"application/json", "vary"=>"Accept,Cookie", "allow"=>"GET, POST, HEAD, OPTIONS", "x-frame-options"=>"SAMEORIGIN", "content-length"=>"44", "via"=>"1.1 pulp3-again.jomitsch.example.com", "connection"=>"close"}
Response body: {"base_path":["This field must be unique."]}
Backtrace:
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/pulp_ansible_client-0.2.0b3.dev01566484022/lib/pulp_ansible_client/api_client.rb:81:in `call_api'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/pulp_ansible_client-0.2.0b3.dev01566484022/lib/pulp_ansible_client/api/distributions_ansible_api.rb:79:in `create_with_http_info'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/pulp_ansible_client-0.2.0b3.dev01566484022/lib/pulp_ansible_client/api/distributions_ansible_api.rb:28:in `create'
/home/vagrant/katello/app/services/katello/pulp3/repository.rb:296:in `create_distribution'
/home/vagrant/katello/app/services/katello/pulp3/repository.rb:285:in `refresh_distributions'
/home/vagrant/katello/app/lib/actions/pulp3/repository/refresh_distribution.rb:18:in `invoke_external_task'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action/polling.rb:83:in `initiate_external_action'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action/polling.rb:18:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action/cancellable.rb:13:in `run'
/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:45:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:538:in `block (3 levels) in execute_run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/katello/app/lib/actions/middleware/execute_if_contents_changed.rb:5:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:31:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/katello/app/lib/actions/middleware/remote_action.rb:16:in `block in run'
/home/vagrant/katello/app/lib/actions/middleware/remote_action.rb:40:in `block in as_remote_user'
/home/vagrant/katello/app/models/katello/concerns/user_extensions.rb:21:in `cp_config'
/home/vagrant/katello/app/lib/actions/middleware/remote_action.rb:27:in `as_cp_user'
/home/vagrant/katello/app/lib/actions/middleware/remote_action.rb:39:in `as_remote_user'
/home/vagrant/katello/app/lib/actions/middleware/remote_action.rb:16:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/rails_executor_wrap.rb:14:in `block in run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-5.2.1/lib/active_support/execution_wrapper.rb:87:in `wrap'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/rails_executor_wrap.rb:13:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action/progress.rb:30:in `with_progress_calculation'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action/progress.rb:16:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:11:in `block in run'
/home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:22:in `with_locale'
/home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:11:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_request_id.rb:15:in `block in run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_request_id.rb:49:in `restore_current_request_id'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_request_id.rb:15:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_timezone.rb:15:in `block in run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_timezone.rb:44:in `restore_curent_timezone'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_timezone.rb:15:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_user.rb:15:in `block in run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_user.rb:44:in `restore_curent_user'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_user.rb:15:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_taxonomies.rb:15:in `block in run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_taxonomies.rb:45:in `restore_current_taxonomies'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/foreman-tasks-0.16.0/app/lib/actions/middleware/keep_current_taxonomies.rb:15:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:26:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:18:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware.rb:31:in `run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/stack.rb:22:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/middleware/world.rb:30:in `execute'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:537:in `block (2 levels) in execute_run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:536:in `catch'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:536:in `block in execute_run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:450:in `block in with_error_handling'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:450:in `catch'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:450:in `with_error_handling'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:531:in `execute_run'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/action.rb:277:in `execute'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:17:in `block (2 levels) in execute'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/execution_plan/steps/abstract.rb:162:in `with_meta_calculation'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:16:in `block in execute'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:30:in `open_action'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:15:in `execute'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/director.rb:43:in `execute'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/executors/parallel/worker.rb:14:in `block in on_message'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/executors.rb:12:in `run_user_code'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/executors/parallel/worker.rb:13:in `on_message'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/context.rb:46:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/executes_context.rb:7:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/actor.rb:98:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/awaits.rb:15:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/sets_results.rb:14:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/actor.rb:46:in `block in on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/actor.rb:57:in `with_backtrace'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.3.0/lib/dynflow/actor.rb:46:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/buffer.rb:38:in `process_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/buffer.rb:31:in `process_envelopes?'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/buffer.rb:20:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/termination.rb:55:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/removes_child.rb:10:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/sets_results.rb:14:in `on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:162:in `process_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:96:in `block in on_envelope'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:119:in `block (2 levels) in schedule_execution'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/synchronization/mutex_lockable_object.rb:41:in `block in synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:116:in `block in schedule_execution'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/serialized_execution.rb:18:in `call'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/serialized_execution.rb:96:in `work'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/serialized_execution.rb:77:in `block in call_job'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

**Comment by sjha4** [comment]
@johnpmitsch : The product name_repo name should be usique in pulp3 db. I have run into this if I reset my katello db and don't reset the pulp3 db..

**Pull request merged**